created --june pride month animation! still needs some work. first PR be kind pls :3

1337isnot1337 opened the pull request:
<!-- Thank you so much for contributing! ❤️ -->

### Description
created the --june pride month animation! this gives a festive little animation whenever it's pride month or if --june flag is used.

### Relevant Links
none
### Screenshots
not really applicable cause there was nothing before but i will add a (choppy) video of it
### Additional context
Still working on it a bit but just wanted to add this for now :3
[Screencast_20240721_164558.webm](https://github.com/user-attachments/assets/3d86daed-6749-49ee-ba26-d742138f764e)

Commit:
created --june pride month animation! still needs some work though. changed cargo toml a bit too and I might have done something wrong.

teohhanhui commented on crates/hyfetch/Cargo.toml:
```diff
@@ -29,6 +30,7 @@ shell-words = { workspace = true, features = ["std"] }
 strum = { workspace = true, features = ["derive", "std"] }
 supports-color = { workspace = true, features = [] }
 tempfile = { workspace = true, features = [] }
+term_size = "0.3.2"
```
This should be removed. We're using `terminal_size` :D

teohhanhui commented on crates/hyfetch/Cargo.toml:
```diff
@@ -15,6 +15,7 @@ ansi_colours = { workspace = true, features = [] }
 anstream = { workspace = true, features = [], optional = true }
 anyhow = { workspace = true, features = ["std"] }
 bpaf = { workspace = true, features = [] }
+clearscreen = "3.0.0"
```
This should be removed. Just use our own `color_util::clear_screen` function. (It's *really* simple btw...)

teohhanhui commented on crates/hyfetch/src/lib.rs:
```diff
@@ -6,3 +6,4 @@ pub mod neofetch_util;
 pub mod presets;
 pub mod types;
 pub mod utils;
+pub mod pride_month;
```
Sort alphabetically.

teohhanhui commented on crates/hyfetch/src/pride_month.rs:
```diff
@@ -0,0 +1,228 @@
+//by 1337isnot1337
```
Remove. We can already see who's the author from the git commits. :smile:

teohhanhui commented on crates/hyfetch/src/pride_month.rs:
```diff
@@ -0,0 +1,228 @@
+//by 1337isnot1337
+
+use crate::{
+    color_util::{color, ForegroundBackground, ToAnsiString},
+    presets::Preset,
+    types,
+};
+use palette::{encoding::Srgb, rgb::Rgb};
+use std::{
+    io::{self, Write},
+    process,
+    sync::atomic::{AtomicBool, Ordering},
+    thread,
+    time::Duration,
+};
+use types::AnsiMode;
+
+use term_size::dimensions;
```
Separate `use` statements into 3 groups: `std`, (extern) crates, `crate`

teohhanhui commented on crates/hyfetch/src/pride_month.rs:
```diff
@@ -0,0 +1,228 @@
+    let notice_end_x = w - 1;
+    let notice_y = h - 1;
+
+    //it is inefficient but I don't know how to iter over values in an enum
```
Something like this:

```rust
use strum::VariantArray as _;

Preset::VARIANTS.iter()
```

teohhanhui commented on crates/hyfetch/src/pride_month.rs:
```diff
@@ -0,0 +1,228 @@
+                //loop over width
+                while x < w {
+                    let idx =
+                    //potential truncation worry but it should be fine
```
It should be fine, so we could remove this comment... :smile_cat:

teohhanhui commented on crates/hyfetch/src/pride_month.rs:
```diff
@@ -0,0 +1,228 @@
+    });
+
+    // the "happy pride month" text that is displayed
+    let text = r".======================================================.
```
Make this a `const` at the file / mod level. Include newlines at beginning and end of the string:

```rust
const TEXT_ASCII: &str = r############################################################"
.======================================================.
| .  .              .__       .     .  .       , .   | |
| |__| _.._ ._   .  [__)._.* _| _   |\/| _ ._ -+-|_  | |
| |  |(_][_)[_)\_|  |   [  |(_](/,  |  |(_)[ ) | [ ) * |
|        |  |  ._|                                     |
'======================================================'
"############################################################;
```

And just do something like this when you use it:

```rust
let text = &TEXT_ASCII[1..TEXT_ASCII.len() - 1];
```

teohhanhui commented on crates/hyfetch/src/pride_month.rs:
```diff
@@ -0,0 +1,228 @@
+                        if (y_text && (text_start_x - border <= x) && (x < text_end_x + border))
+                            || (y == notice_y && notice_start_x - 1 <= x && x < notice_end_x + 1)
+                        {
+                            //buf += c.overlay(black, 0.5).to_ansi_rgb(foreground=False)
+                            buf += // <--- && ^^^^^^ need to print a darker background somehow
```
~~https://docs.rs/palette/latest/palette/blend/trait.Blend.html#tymethod.overlay~~

^^

EDIT: I think this is what we need: https://github.com/Ogeon/palette/blob/85502508b28d5b1a69a67b734ee3d896bfa9ddae/palette/src/blend.rs#L16-L30

https://docs.rs/palette/latest/palette/blend/struct.Equations.html

EDIT 2: mm... I think this is exactly what we need lol: https://docs.rs/palette/latest/palette/blend/trait.Compose.html#tymethod.over

Looking at the code, yup: https://github.com/Ogeon/palette/blob/85502508b28d5b1a69a67b734ee3d896bfa9ddae/palette/src/blend/compose.rs#L53-L61

Commit:
FIxed most bugs, should be working fine now?

1337isnot1337 commented:
These bugs/issues were fixed:

- moved from term_size to terminal_size
- sorted lib.rs
- *tried* to rearrange use statements
- used strum to iterate over enums (did not know this, thank you for informing me about it!)
- removed unneeded comment
- make the text a constant and reduced redundancy
- the text will now have a darker, transparent background so that the colors are more visible

Hope that this is better! anything else I need @teohhanhui?

teohhanhui committed:
Clean up pride_month.rs

teohhanhui commented:
@1337isnot1337 Thank you! :rocket:

1337isnot1337 commented:
> @1337isnot1337 Thank you! 🚀

@teohhanhui No problem! Very fun first pull request. Thank you so much :)
